### PR TITLE
Return error when a connection observer detect a disconnection

### PR DESCRIPTION
**Pull request by DrAma999**
Sometimes can be useful to detect if a disconnection is unfortunate and, based on the error received, apply some reconnection logic or avoid the reconnection completely, or just to log the error and send it to a server.
To do that I've applied some modification to the `ConnectionObserver` protocol in the disconnected method.
`    public func disconnected(from peripheral: Bluejay.Peripheral, with error: Error?)
`
This method just forwards the error received from the `CBCentralManager ` `    public func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?)
` callback.

### CODE REVIEW

**Comment by JeremyChiang** [comment]
Thanks for the PR. I will be reviewing this now. Before I start, I am going to change base from `master` to `v0.7.0`.

File: Bluejay/Bluejay/Bluejay.swift [hunk 1]
```diff
@@ -1008,7 +1008,7 @@ extension Bluejay: CBCentralManagerDelegate {
         }
         
         for observer in observers {
-            observer.weakReference?.disconnected(from: disconnectedPeripheral)
+            observer.weakReference?.disconnected(from: disconnectedPeripheral, with: error)
```
Comment (JeremyChiang) [inline Bluejay/Bluejay/Bluejay.swift]: Now that we include the error, there is an even higher chance that people will be tempted and led to writing some kind reconnection logic in their `disconnected` event callbacks. That wouldn't work as it is right now, because at this point we haven't reached the `disconnectCleanUp` yet, so any new `Connection` operation being added to the queue during the `disconnect` event will fail immediately as Bluejay will still think we're connected.

We may have to move this entire for-loop block after line 1053, after niling out `connectingPeripheral` and `connectedPeripheral`. But there are several problems with that as well:
1. Unclear if moving the notification of observers will have any negative impact, but it certainly does change the whole procedures here as the order of operations is really important. We will have to be very careful as it could break some apps in production.
- One thing that will happen for sure if we move the for-loop clause is that you will get a `notConnected` error in your in-flight and already queued operations, such as any recently called reads or writes, and then your register observers will get the `disconnected` event last. That's not necessarily wrong, but it's a difference.
2. If we are expecting users to write their own reconnection logic, then we need to force them to handle Bluejay's auto reconnection at this step as well - so in the disconnection event, the user needs to be able to:
  - Check whether Bluejay will automatically reconnect, if so, maybe no need to write your own reconnection
  - Ability to cancel Bluejay's automatic reconnect, so users can either use their own reconnection code or just ask Bluejay to not try to reconnect
  - Maybe all of this means a 3rd parameter for the `disconnected` delegation of a callback type with a default value (so it's optional) that will then be invoked inside `disconnectCleanUp`. Not sure yet, will have to think a bit more.

Comment (DrAma999) [reply Bluejay/Bluejay/Bluejay.swift]: Sorry, for the first file totally useless.
About your considerations, yep it's all true. I also had this issue after disconnecting and trying to reconnect on another device same issue as the user reported in [#168](https://github.com/steamclock/bluejay/issues/168). 
I will look into that and try to propose a solution.

File: Bluejay/Bluejay/ConnectionObserver.swift [hunk 2]
```diff
@@ -30,13 +30,18 @@ public protocol ConnectionObserver: class {
...
+    /// Called before Blujay tries to reconnect from an accidental disconnection from a peirpheral
+    func shouldReconnect(to: Peripheral) -> Bool
 }
 
 /// Slightly less gross way to make the ConnectionObserver protocol's functions optional.
 extension ConnectionObserver {
     public func bluetoothAvailable(_ available: Bool) {}
     public func connected(to peripheral: Peripheral) {}
     public func disconnected(from peripheral: Peripheral, with error: Error?) {}
+    public func shouldReconnect(to: Peripheral) -> Bool { return true }
```
[the remark this answers is not in the document]
Comment (DrAma999) [reply Bluejay/Bluejay/ConnectionObserver.swift]: Hello @JeremyChiang, I'm sorry but the work isn't done yet. I thought it was more simple, but there are a lot of  corner cases that still must be handled. I still didn't test it yet and probably there are still some "ideas" that I left mistakenly in the code. I was really busy at work and since the last commit I didn't had time to work again on this issue.
But I think it could be a great time to clarify what's on my mind.
The first startegy I tried to apply was on connection observers, unfortunately this could never work, primary because if there is more than one connection observer active as listeners is impossible to decide who must decide if it should reconnect or not. What you are still seeing in the code is a misprint about this old approach that I must remember to remove.
The second strategy I applied is the one with the `reconnectionHandler`, this one is set on Bluejay instance itself. I'm pretty confident that this should work, but I have some concern about cancellation: I'm not totally sure that if I cancel a connection `CBCentralManager` will notify `didDisconnectPeripheral` or `didFailToConnect` by returning an `Error` and making the user with the `reconnectionHandler ` apply its logic as such an event occur. As I said I still need to test it.
About the `BluejayConfiguration` I didn't know about the #165 , totally correct about what you say at point 3, but point 2 I think is just a matter of taste, I applied this patter since also in Apple frameworks is appearing more often, for instance in `UIGraphicsImageRenderer`, you can pass a format that specifies some options and Apple provides some defaults values as well, but no big deal about.
I hope to start testing it next week.

**Comment by DrAma999** [comment]
I removed the `BluejayConfiguration` struct and added the `shouldReconnectHandler` to Bluejay initialization.
I've been testing it and it seems to work pretty fine.
This new proposal provides a `shouldReconnectHandler`of type `typealias ShouldReconnectHandler = (Peripheral, Error?) -> Bool` as a property of Bluejay this is a block of that as default returns true and basically overrides the auto-connect behavior.
Now users can set this callback and implement their own reconnection logic since this handler is called right after a disconnection is detected.

**Comment by DrAma999** [comment]
> Please check out my new PR that may replace this: #170

Wow... that is great. I think I'm going to start your PR next week.

**Comment by JeremyChiang** [comment]
Closing this in favour of the more comprehensive and tested PR: https://github.com/steamclock/bluejay/pull/170
